CONTRIBUTING.md

```diff
@@ -13,14 +13,77 @@ Please fill out either the individual or corporate Contributor License Agreement
+
+### Where to Submit Your Patch
+
+These folders are automatically generated. You will need to submit a patch to the upstream kubernetes repo [kubernetes](https://github.com/kubernetes/kubernetes) or the OpenAPI generator repo [openapi-generator](https://github.com/OpenAPITools/openapi-generator). This contains:
+- [kubernetes/client](kubernetes/client)
+- [kubernetes/test](kubernetes/test)
+- [kubernetes/docs](kubernetes/docs).
+
+These folders contain developer written codes in the main repo and the patches should be submitted here:
+- [kubernetes/e2e_test](kubernetes/e2e_test)
+- [kubernetes/utils](kubernetes/utils)
+- [examples](examples)
+- [scripts](scripts).
+
+The base repo contains developer written codes only. The patches should be sent to the base repo instead. The scope covers these repos and symbolic links in the main repo:
+- kubernetes/base
+- kubernetes/config
+- kubernetes/stream
+- kubernetes/watch.
+
 ### Contributing A Patch
 
 1. Submit an issue describing your proposed change to the repo in question.
-1. The [repo owners](OWNERS) will respond to your issue promptly.
-1. If your proposed change is accepted, and you haven't already done so, sign a Contributor License Agreement (see details above).
-1. Fork the desired repo, develop and test your code changes.
-1. Submit a pull request.
+2. The [repo owners](OWNERS) will respond to your issue promptly.
+3. If your proposed change is accepted, and you haven't already done so, sign a Contributor License Agreement (see details above).
+4. Fork the desired repo, develop and test your code changes. Add a test if possible.
+5. Submit a pull request.
 
 ### Adding Dependencies
 
-If your patch depends on new packages, add those packages to [requirements.txt](requirements.txt) and [setup.py](setup.py).
+If your patch depends on new packages, add those packages to [requirements.txt](requirements.txt) and/or [setup.py](setup.py). If these package are for testing only, add those to [test-requirements.txt](test-requirements.txt).
+
+### Commits
+
+Generally we would like to see one commit per pull request. However, if the pull request is reasonably large, the PR can be divided into several commits that make logical sense. The commit message should be clear and indicative of the aim of the fix.
+
+If you have several commits in a pull request and have been asked to squash your commits, please use ```git reset --soft HEAD~N_COMMITS``` and commit again to make your PR a single commit.
+
+### Windows Developers
+
+The symbolic links contained in this repo does not work for Windows operating systems. If you are a Windows developer, please run the [fix](scripts/windows-setup-fix.bat) inside the scripts folder or manually copy the content of the [kubernetes/base](https://github.com/kubernetes-client/python-base) folder into the [kubernetes](kubernetes) folder.
```

Comment:
minor nitpick: "does not work" should be "do not work".

[kubernetes/base](https://github.com/kubernetes-client/python-base) -> [kubernetes-client/python-base](https://github.com/kubernetes-client/python-base)

Reply:
Thanks! All other comments look good to me, but the kubernetes/base folder is what the Windows development sees after they clone the repo. This part is unchanged.